### PR TITLE
GDTCORFlatFileStorage: keep not expired events when expired batch removed

### PR DESCRIPTION
**Pull request by maksymmalyhin**
Fixes the following issue. It it possible that a batch expiration time is shorter than an event expiration time. With the old implementation in this case when the batch expires it will lead to removing all the events in it disregarding the event expiration time. 

The proposed fix:
- check batches for expiration first
- remove expired batches and move events from them back to the main location (unconditionally) 
- check and remove expired events from the main location (expired events from expired batches will also be there)

### CODE REVIEW

File: GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m [hunk 1]
```diff
@@ -384,45 +341,46 @@ - (void)hasEventsForTarget:(GDTCORTarget)target onComplete:(void (^)(BOOL hasEve
...
-        [pathsToDelete addObject:[eventDataPath stringByAppendingPathComponent:path]];
-      }
-    }
 
-    // TODO: Events from expired batches with not expired events must be moved back to queue to
-    // avoid data loss.
     // TODO: Storage may not have enough context to remove batches because a batch may be being
     // uploaded but the storage has not context of it.
+
+    // Find expired batches and remove them moving the events back to the main storage.
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m:351]: Do you mean 

"Find expired batches and move their events back to the main storage."

?

Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m:351]: yes, exactly what I mean, thank you! Sorry, I corrected the comment a couple times and ended up with a weird version.

File: GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m [hunk 2]
```diff
@@ -539,6 +497,53 @@ - (BOOL)moveContentsOfDirectoryAtPath:(NSString *)sourcePath
...
+    if (deleteEvents) {
+      removeBatchDir(batchDirPath);
+    } else {
+      NSString *batchDirName = [batchDirPath lastPathComponent];
+      NSDictionary<NSString *, id> *components = [self batchComponentsFromFilename:batchDirName];
+      NSNumber *target = components[kGDTCORBatchComponentsTargetKey];
+      NSString *destinationPath = [[GDTCORFlatFileStorage eventDataStoragePath]
+          stringByAppendingPathComponent:target.stringValue];
+
+      // `- [NSFileManager moveItemAtPath:toPath:error:] method fails if an item by the
```
Comment (mikehaney24) [inline GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m:530]: Unbalanced backtick

File: GoogleDataTransport/GDTCORTests/Unit/GDTCORFlatFileStorageTest.m [hunk 3]
```diff
@@ -783,8 +790,27 @@ - (void)testBatchIDWithTarget {
   [self waitForExpectations:@[ expectation ] timeout:10];
 }
 
+- (void)testBatchIDsForTarget {
+  __auto_type expectedBatch = [self generateAndBatchEvents];
```
Comment (mikehaney24) [inline GoogleDataTransport/GDTCORTests/Unit/GDTCORFlatFileStorageTest.m:794]: I assume this compiles in Xcode 10?

File: GoogleDataTransport/GDTCORTests/Unit/GDTCORFlatFileStorageTest.m [hunk 4]
```diff
@@ -823,78 +849,88 @@ - (void)testCheckEventExpiration {
...
+- (void)testCheckForExpirations_WhenBatchWithNotExpiredEventsExpires {
+  NSTimeInterval batchExpiresIn = 0.5;
+  // 0.1. Generate and batch events
+  __auto_type generatedBatch = [self generateAndBatchEventsExpiringIn:1000
+                                                      batchExpiringIn:batchExpiresIn];
+  NSNumber *generatedBatchID = [[generatedBatch allKeys] firstObject];
+  NSSet<GDTCOREvent *> *generatedEvents = generatedBatch[generatedBatchID];
+  // 0.2. Wait for batch expiration.
+  [[NSRunLoop currentRunLoop]
+      runUntilDate:[NSDate dateWithTimeIntervalSinceNow:batchExpiresIn + 0.5]];
```
Comment (mikehaney24) [inline GoogleDataTransport/GDTCORTests/Unit/GDTCORFlatFileStorageTest.m:861]: Maybe move the 0.5 to a variable?

File: GoogleDataTransport/GDTCORTests/Unit/GDTCORFlatFileStorageTest.m (hunk 3 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORTests/Unit/GDTCORFlatFileStorageTest.m:794]: Yes, fortunately it does.

File: GoogleDataTransport/GDTCORTests/Unit/GDTCORFlatFileStorageTest.m (hunk 4 above)
Comment (maksymmalyhin) [reply GoogleDataTransport/GDTCORTests/Unit/GDTCORFlatFileStorageTest.m]: Seems it should work without additional delays, so I removed it.

**Comment by maksymmalyhin** [comment]
@mikehaney24 Thank you for the review.

I like the idea with a block either passed to `dispatch_async` or just executed. If you don't mind I'll consider it for the future changes - I still need to wrap my head around how completion block semantic will look like in this case.

I tried to put the answer to your question to the PR description (I should've done it initially). If I understand the problem correctly - the expired events from the expired batches will be also removed with the proposed implementation.

**Pull request merged**